nixos/modules/services/networking/rustical.nix

```
@@ -0,0 +1,170 @@
+      '';
+      example = {
+        RUSTICAL_HTTP__PORT = 4000;
+      };
+    };
+    dataDir = mkOption {
+      type = path;
+      default = "/var/lib/rustical";
+      description = "The state directory for Rustical";
+    };
+    openFirewall = lib.mkOption {
+      type = bool;
+      default = false;
+      description = "Open ports in the firewall for the Rustical.";
+    };
+    user = mkOption {
+      type = str;
+      default = "rustical";
+      description = "User under which the rustical service runs.";
+    };
+    group = mkOption {
+      type = str;
+      default = "rustical";
+      description = "Group under which the rustical service runs.";
+    };
+  };
+
+  config = mkIf cfg.enable {
+    services.rustical.settings = {
+      # We assume dataDir is used, so we require this value,
+      # therefore no mkDefault
+      data_store.sqlite.db_url = "${cfg.dataDir}/db.sqlite3";
+      http.host = mkDefault "0.0.0.0";
+      frontend = {
+        enabled = mkDefault true;
+        allow_password_login = mkDefault true;
+      };
+      tracing.opentelemetry = mkDefault false;
+      dav_push.enabled = mkDefault true;
+      nextcloud_login.enabled = mkDefault true;
+    };
+
+    networking.firewall = mkIf cfg.openFirewall {
+      allowedTCPPorts = [ cfg.settings.http.port ];
+    };
+
+    users = {
+      users.${cfg.user} = {
+        isSystemUser = true;
+        group = cfg.group;
+      };
+      groups.${cfg.group} = { };
+    };
+
+    systemd = {
+      services.rustical = {
+        description = "A CalDAV/CardDAV server";
+        after = [ "network.target" ];
+        wantedBy = [ "multi-user.target" ];
+        environment = cfg.environment;
+        serviceConfig = {
+          Type = "simple";
+          ExecStart = "${getExe cfg.package} --config-file ${settingsFormat.generate "config.toml" cfg.settings}";
+          EnvironmentFile = mkIf (cfg.environmentFile != null) cfg.environmentFile;
+          StateDirectory = mkIf (cfg.dataDir == "/var/lib/rustical") "rustical";
+          User = cfg.user;
+          Group = cfg.group;
+          Restart = "on-failure";
+          # Hardening
+          SystemCallFilter = "@system-service";
+          SystemCallErrorNumber = "EPERM";
+          ProtectSystem = "strict";
+          ProtectHome = "yes";
+          ProtectKernelTunables = "yes";
+          ProtectKernelModules = "yes";
+          ProtectControlGroups = "yes";
+          ProtectProc = "invisible";
+          ProcSubset = "pid";
+          CapabilityBoundingSet = "";
+          AmbientCapabilities = "";
+          PrivateDevices = "yes";
+          PrivateTmp = "yes";
+          NoNewPrivileges = "yes";
+          ReadWritePaths = cfg.dataDir;
+          ProtectHostname = "yes";
+          ProtectClock = "yes";
+        };
+      };
+      tmpfiles.settings.rustical."${cfg.dataDir}".d = mkIf (cfg.dataDir != "/var/lib/rustical") {
+        inherit (cfg) user group;
+
+        mode = "0755";
```

It's probably no big deal if you apply the other hardening suggestion with `UMask`, but I assume you don't actually want random users on the system to be able to read the database and other state of the application?

```suggestion
        mode = "0750";
```